### DISCUSSION TITLE
Does Pydantic natively support schema validation of JSON schemas?

### DISCUSSION THREAD

**Discussion by Lance-Drane** [pydantic/pydantic#5135] [Question] [answered]
I'm currently using both jsonschema and Pydantic as libraries in a library I'm working on. However, I'm only using jsonschema like this:

```python

from typing import Any, Dict, Optional

from pydantic import BaseModel, validator

class Config(BaseModel):
  argument_schema: Optional[Dict[str, Any]] = None
   
  @validator("argument_schema")
  def _valid_json_schema(cls, v):
    # this code will only get called once per application lifecycle
    if v is not None:
      from jsonschema import Draft202012Validator as SchemaValidator
      from jsonschema import SchemaError

      try:
        SchemaValidator.check_schema(v)
      # TODO seems to only catch a single error - jsonschema iterates over a Generator to raise multiple errors
      except SchemaError as e:
        raise ValueError(f"Schema config error: {e.message}:\n    {v}")
    return v
```

This code mostly works fine for me, but I would ideally like to accomplish metaschema validation using Pydantic, so I can reduce the number of dependencies in my library. I dug through [the code in json_schema.py](https://github.com/pydantic/pydantic/blob/main/pydantic/json_schema.py) but I couldn't find any functionality which would suggest validation of a metaschema against a schema. (As far as I can tell, nothing is being done with`$schema` and there's otherwise no definition of using a specific format, and nothing in json_schema.py is imported into `__init__.py` for the sake of public usage. )

So my questions:
1) Does Pydantic handle metaschema validation (without me having to instatiate a BaseModel Schema class myself), and I'm overlooking how it does it?
2) If Pydantic does not handle this functionality, is there any interest in doing so?

Thanks for reading!

**Comment by dmontagu** [comment, marked as answer]
Currently, `pydantic` does nothing to validate JSON schema whatsoever — either that a JSON schema is valid, _or_ that a JSON object matches a JSON schema. The extent of pydantic's JSON schema integration today is to generate JSON schema for various types, and I believe was originally added by @tiangolo for the purposes of FastAPI.

I think at this point in time, validating JSON schema (either instances against a schema, or schemas against a metaschema) is out of scope for pydantic proper, and I would encourage you to take an approach like you've taken using a library like `jsonschema` to implement a pydantic validator if you want to perform actual validation.

That said, I could see it being reasonable to add a JSON schema validator type to the nascent `pydantic_extra_types` package. I think even if that was done though, we'd likely take the approach of relying on a library like `jsonschema` to maintain the actual validation logic.

**Reply by Lance-Drane** [reply]
Thanks for the quick response! After looking through both libraries there does appear to be a clear separation of concerns, though one thought I had would be to have the option to leverage Pydantic’s validators for specific [formats.](https://github.com/python-jsonschema/jsonschema/blob/main/jsonschema/_format.py)

**Comment by samuelcolvin** [comment]
If your JSONSchemas are static, you can use https://docs.pydantic.dev/datamodel_code_generator/ to convert JSONSchema to python code representing pydantic models.

In future, there's the possibility that someone could write a converter from JSONSchema to pydantic-core schema.

**Comment by fritz-astronomer** [comment]
+1 - would be a cool addition to `pydantic_extra_types`, to have a `JsonSchema` type field similar to how there's currently a `JSON` type field. 

It might be a niche use case, but it makes sense for an object that can take a JSON Schema during instantiation as a str or dict. 
Given that JSONSchema has a relatively set structure - it'd help with introspection, validation, and type hinting. 

I'm surprised the `jsonschema` package doesn't have anything for type hints

**Comment by akshaylive** [comment]
I developed [jsonschema-pydantic-converter](https://github.com/akshaylive/jsonschema-pydantic-converter) to handle a related use case. It represents JSON schemas as dynamic types at runtime rather than generating static code.

This is particularly useful when you need Pydantic's validation and serialization for dynamic schemas, or when integrating with libraries like LangChain that require Pydantic types for tool definitions but where materializing code isn't feasible.
